src/nnet3/nnet-optimize.h

```diff
@@ -76,18 +81,67 @@ void Optimize(const NnetOptimizeOptions &config,
               const ComputationRequest &request,
               NnetComputation *computation);
 
+// Hash function for ComputationRequest. It converts
+// ComputationRequest to hash code by looking at input
+// and output IoSpecifications vectors.
+struct ComputationRequestHasher {
+  size_t operator()(const ComputationRequest *cr) const {
+    size_t ans = 0;
+    std::vector<IoSpecification>::const_iterator itr = cr->inputs.begin(),
+                                                 end = cr->inputs.end();
+    for (; itr != end; ++itr) {
+      ans += IoSpecificationToInt(*itr);
+    }
+    itr = cr->outputs.begin();
+    end = cr->outputs.end();
+    for (; itr != end; ++itr) {
+      ans += IoSpecificationToInt(*itr);
+    }
+    return ans;
+  }
+ private:
+  size_t IoSpecificationToInt(const IoSpecification& spec) const {
+    size_t ans, slen = spec.name.length();
+    const char *c = spec.name.c_str(), *send = c + slen;
```

Comment:
for this part of the code you can use the StringHasher() object, e.g.
StringHasher string_hasher;
size_t ans = string_hasher(spec.name);
BTW, your code was using an unitialized variable (ans).  You should run your code on valgrind to see if there are other instances of this.